resharper/src/resharper-unity/Host/Features/Foldings/ShaderLab/ShaderLabCodeFoldingProcessor.cs

```diff
@@ -30,28 +30,20 @@ public override void VisitNode(ITreeNode node, IHighlightingConsumer consumer)
                 ShaderLabTokenType.MULTI_LINE_COMMENT, ShaderLabTokenType.NEW_LINE);
         }
 
+#pragma warning disable 672
+        // Obsolete warning tells us to also implement VisitTexturePropertyValueNode
```

Comment:
Why are we ignoring the warning? Is `VisitBlockValueNode` not really obsolete?

Reply:
It's what happens when an element is used in multiple interfaces - the `Visit{Whatever}Node` gets the `Obsolete` attribute to warn that the overload won't get called for certain element types (and it lists which types). It's not obsolete, but it's a way of getting a warning into the compiler. The overload is still called for elements that only implement the `IBlockValue` interface.

I don't really like it (I'd prefer to compile with warnings as errors on, ideally), but I think I want to keep the multiple interfaces, at least for now. It's useful to be able to treat `TexturePropertyValue` as both a property value (used in `Properties`) and a block value (has `{` and `}`, useful for folding. It might be that as we use the AST for more things, it doesn't make as much sense, and then we can perhaps reconsider. Perhaps `IBlockValue` should move out of the parser, and be added by hand in partial classes.

And I checked the C# parser, and it's got loads of them, too :)